lib/go-atscfg/parentdotconfig.go

```diff
@@ -331,13 +330,268 @@ func MakeParentDotConfig(
+	if serverIsLastTier {
+		orgURI, err := GetOriginURI(ds.OriginFQDN) // TODO pass, instead of calling again
+		if err != nil {
+			return nil, nil, err
+		}
+		return []string{orgURI.Host}, nil, nil
+	}
+
+	svNode := tc.TopologyNode{}
+	for _, node := range topology.Nodes {
+		if node.Cachegroup == server.Cachegroup {
+			svNode = node
+			break
+		}
+	}
+	if svNode.Cachegroup == "" {
+		return nil, nil, errors.New("This server '" + server.HostName + "' not in DS " + string(ds.Name) + " topology, skipping")
+	}
+
+	if len(svNode.Parents) == 0 {
+		return nil, nil, errors.New("DS " + string(ds.Name) + " topology '" + ds.Topology + "' is last tier, but NonLastTier called! Should never happen")
+	}
+	if numParents := len(svNode.Parents); numParents > 2 {
+		log.Errorln("DS " + string(ds.Name) + " topology '" + ds.Topology + "' has " + strconv.Itoa(numParents) + " parent nodes, but Apache Traffic Server only supports Primary and Secondary (2) lists of parents. CacheGroup nodes after the first 2 will be ignored!")
+	}
+	if len(topology.Nodes) <= svNode.Parents[0] {
+		return nil, nil, errors.New("DS " + string(ds.Name) + " topology '" + ds.Topology + "' node parent " + strconv.Itoa(svNode.Parents[0]) + " greater than number of topology nodes " + strconv.Itoa(len(topology.Nodes)) + ". Cannot create parents!")
+	}
+	if len(svNode.Parents) > 1 && len(topology.Nodes) <= svNode.Parents[1] {
+		log.Errorln("DS " + string(ds.Name) + " topology '" + ds.Topology + "' node secondary parent " + strconv.Itoa(svNode.Parents[1]) + " greater than number of topology nodes " + strconv.Itoa(len(topology.Nodes)) + ". Secondary parent will be ignored!")
+	}
+
+	parentCG := topology.Nodes[svNode.Parents[0]].Cachegroup
+	secondaryParentCG := ""
+	if len(svNode.Parents) > 1 && len(topology.Nodes) > svNode.Parents[1] {
+		secondaryParentCG = topology.Nodes[svNode.Parents[1]].Cachegroup
+	}
+
+	if parentCG == "" {
+		return nil, nil, errors.New("Server '" + server.HostName + "' DS " + string(ds.Name) + " topology '" + ds.Topology + "' cachegroup '" + server.Cachegroup + "' topology node parent " + strconv.Itoa(svNode.Parents[0]) + " is not in the topology!")
+	}
+
+	parentStrs := []string{}
+	secondaryParentStrs := []string{}
+	for _, sv := range servers {
+		if tc.CacheType(sv.Type) != tc.CacheTypeEdge && tc.CacheType(sv.Type) != tc.CacheTypeMid && sv.Type != tc.OriginTypeName {
+			continue // only consider edges, mids, and origins in the CacheGroup.
+		}
+		if !HasRequiredCapabilities(serverCapabilities[sv.ID], ds.RequiredCapabilities) {
+			continue
+		}
+		if sv.Cachegroup == parentCG {
+			parentStr := serverParentStr(sv, parentConfigParams)
+			if parentStr != "" { // will be empty if server is not_a_parent (possibly other reasons)
+				parentStrs = append(parentStrs, parentStr)
+			}
+		}
+		if sv.Cachegroup == secondaryParentCG {
+			secondaryParentStrs = append(secondaryParentStrs, serverParentStr(sv, parentConfigParams))
+		}
```

Comment:
If the immediate parent cachegroup has no servers in it `GetTopologyParents()`, returns empty arrays, even if there are servers in further ancestor topology nodes. Should it keep going until it finds a topology node with a non-empty cachegroup?

Reply:
On second thought, maybe we should perform validation to make sure a topology node's cachegroup contains at least 1 server.

Reply:
(That validation would occur in TO and is outside the scope of this PR)

Reply:
I've been trying to log errors on every possible bad data the config gen can receive.
I'd agree with doing the validation in TO, but also logging an error in the config gen. I'll make it do that

Reply:
Ok, I made it log the error